tests/python_package_test/test_engine.py

```diff
@@ -3224,3 +3224,34 @@ def test_force_split_with_feature_fraction(tmp_path):
     for tree in tree_info:
         tree_structure = tree["tree_structure"]
         assert tree_structure['split_feature'] == 0
+
+
+def test_record_evaluation_with_train():
+    X, y = make_synthetic_regression()
+    ds = lgb.Dataset(X, y)
+    eval_result = {}
+    callbacks = [lgb.record_evaluation(eval_result)]
+    params = {'objective': 'l2', 'num_leaves': 3}
+    num_boost_round = 5
+    bst = lgb.train(params, ds, num_boost_round=num_boost_round, valid_sets=[ds], callbacks=callbacks)
+    assert list(eval_result.keys()) == ['training']
+    train_mses = []
+    for i in range(num_boost_round):
+        pred = bst.predict(X, num_iteration=i + 1)
+        mse = mean_squared_error(y, pred)
+        train_mses.append(mse)
+    np.testing.assert_allclose(eval_result['training']['l2'], train_mses)
+
+
+def test_record_evaluation_with_cv():
+    X, y = make_synthetic_regression()
+    ds = lgb.Dataset(X, y)
+    eval_result = {}
+    callbacks = [lgb.record_evaluation(eval_result)]
+    metrics = ['l2', 'rmse']
+    params = {'objective': 'l2', 'num_leaves': 3, 'metric': metrics}
+    cv_hist = lgb.cv(params, ds, num_boost_round=5, stratified=False, callbacks=callbacks, eval_train_metric=True)
```

Comment:
Proposed solution crashes if you set `eval_train_metric=False` in this test case.
```
ValueError                                Traceback (most recent call last)
<ipython-input-5-e50e682e859d> in <module>
----> 1 cv_hist = lgb.cv(params, data_train, num_boost_round=5, stratified=False, callbacks=callbacks, eval_train_metric=False)

D:\Miniconda3\lib\site-packages\lightgbm\engine.py in cv(params, train_set, num_boost_round, folds, nfold, stratified, shuffle, metrics, fobj, feval, init_model, feature_name, categorical_feature, early_stopping_rounds, fpreproc, verbose_eval, show_stdv, seed, callbacks, eval_train_metric, return_cvbooster)
    645         try:
    646             for cb in callbacks_after_iter:
--> 647                 cb(callback.CallbackEnv(model=cvfolds,
    648                                         params=params,
    649                                         iteration=i,

D:\Miniconda3\lib\site-packages\lightgbm\callback.py in _callback(env)
    141     def _callback(env: CallbackEnv) -> None:
    142         if not eval_result:
--> 143             _init(env)
    144         for item in env.evaluation_result_list:
    145             if len(item) == 4:

D:\Miniconda3\lib\site-packages\lightgbm\callback.py in _init(env)
    135             else:  # cv
    136                 data_eval = item[1]
--> 137                 data_name, eval_name = data_eval.split()
    138             eval_result.setdefault(data_name, collections.OrderedDict())
    139             eval_result[data_name].setdefault(eval_name, [])

ValueError: not enough values to unpack (expected 2, got 1)
```

item in `env.evaluation_result_list` will look like
```
('cv_agg', 'l2', 0.077931728049798, False, 0.00858679769235092)
```

Reply:
I fixed this for this callback by setting the first level key to `valid` even when there's no train metric being computed to preserve the structure but maybe it'd be better if the second element from this result list was 'valid l2' to be consistent. WDYT?

Reply:
By that I mean removing the if in https://github.com/microsoft/LightGBM/blob/a06fadfb7ac3fdb26da3d4afc061a8b976070c50/python-package/lightgbm/engine.py#L370-L373 and set `key = f"{one_line[0]} {one_line[1]}"` always.

Reply:
> ... but maybe it'd be better if the second element from this result list was 'valid l2' to be consistent.

Yeah, I support this.

Reply:
changed in 914818c4bc70142c159d9865e0b1583486947c8d

Reply:
I'm going to mark this PR as `breaking` due to this change.